// @nrwl/angular/src/generators/setup-ssr/lib/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const devkit_1 = require("@nrwl/devkit");
function normalizeOptions(tree, options) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    return {
        project: (_a = options.project) !== null && _a !== void 0 ? _a : (0, devkit_1.readWorkspaceConfiguration)(tree).defaultProject,
        appId: (_b = options.appId) !== null && _b !== void 0 ? _b : 'serverApp',
        main: (_c = options.main) !== null && _c !== void 0 ? _c : 'main.server.ts',
        serverFileName: (_d = options.serverFileName) !== null && _d !== void 0 ? _d : 'server.ts',
        serverPort: (_e = options.serverPort) !== null && _e !== void 0 ? _e : 4000,
        rootModuleFileName: (_f = options.rootModuleFileName) !== null && _f !== void 0 ? _f : 'app.server.module.ts',
        rootModuleClassName: (_g = options.rootModuleClassName) !== null && _g !== void 0 ? _g : 'AppServerModule',
        skipFormat: (_h = options.skipFormat) !== null && _h !== void 0 ? _h : false,
    };
}
exports.normalizeOptions = normalizeOptions;
